test/types/config_types.ts

```diff
@@ -63,3 +63,51 @@ chart4.data.datasets.push({
   type: 'line',
   data: [1, 2, 3]
 });
+
+const chart5 = new Chart('chart', {
+  data: {
+    labels: ['1', '2', '3'],
+    datasets: [{
+      type: 'line',
+      data: [{
+        category: '1',
+        value: 1
+      }]
+    }]
+  }
+});
+
+chart5.data.datasets.push({
+  type: 'line',
+  data: [{
+    category: '1',
+    value: 1
+  }]
+});
+
+chart5.config.data.datasets.push({
+  type: 'line',
+  data: [{
+    category: '2',
+    value: 2
+  }]
+});
+
+const dataset = chart5.data.datasets[0];
+// number | Point | [number, number] | BubbleDataPoint | { category: string, value: number }
+const datapoint = dataset.data[0];
+
+if (typeof datapoint === 'object' && 'category' in datapoint) {
+  // datapoint: { category: string, value: number }
+  datapoint.value = 2;
+}
+
+if (dataset.type === 'line') {
+  // number | Point | { category: string, value: number }
+  const lineDatapoint = dataset.data[0];
```

Comment:
@LeeLenaleee 

Above, we don't know the dataset type, so the datapoint can be ` number | Point | [number, number] | BubbleDataPoint | { category: string, value: number }`. Here we know the dataset type, so here datapoint type is `number | Point | { category: string, value: number }`.